### PR TITLE
CM-25773 - Support TF Plan scans

### PR DESCRIPTION
**Pull request by EfratIsrael**

### CODE REVIEW

**Review by doratias18 — COMMENTED** [review]
Add more tests

**Comment by doratias18** [comment]
@artem-fedorov @MichalBor @MarshalX @avishaiamiel please review

File: cycode/cli/code_scanner.py
Comment (MichalBor) [inline cycode/cli/code_scanner.py:342]: scan_disk_files is relevant only for scan type `path`, is it intended? we don't  want to support it also in `repository` and `commit_range`?

File: cycode/cli/code_scanner.py
Comment (MichalBor) [inline cycode/cli/code_scanner.py:346]: do we want to soft fail and exit when failing to generate tfplan?

File: cycode/cli/code_scanner.py
Comment (MichalBor) [reply cycode/cli/code_scanner.py:342]: why we do the logic here and not in the server, its not something that should be done also as part of repository/pr scan?

File: cycode/cli/helpers/tf_content_generator.py
Comment (MichalBor) [inline cycode/cli/helpers/tf_content_generator.py:27]: `json.dumps(value)` works for you if its a dict? the inner fields format will be like json - `key: value` instead of `key = value"

File: cycode/cli/code_scanner.py
Comment (MichalBor) [inline cycode/cli/code_scanner.py:1124]: maybe check `terraform version` key exists as well?

File: cycode/cli/helpers/tf_content_generator.py
Comment (MichalBor) [inline cycode/cli/helpers/tf_content_generator.py:17]: is there `after` in case of resource deletion?

Comment (MichalBor) [reply cycode/cli/helpers/tf_content_generator.py:17]: please change all to safe get

File: cycode/cli/utils/file_utils.py
Comment (MichalBor) [inline cycode/cli/utils/file_utils.py:4]: I would move it to path_utils

File: cycode/cli/utils/file_utils.py
Comment (MichalBor) [inline cycode/cli/utils/file_utils.py:5]: nice :)

File: cycode/cli/helpers/tf_content_generator.py
Comment (MichalBor) [inline cycode/cli/helpers/tf_content_generator.py:22]: I would put it in ChangeResource as magic method `__str__ ` or `__repr__`

File: tests/cli/helpers/test_tf_content_generator.py
Comment (MichalBor) [inline tests/cli/helpers/test_tf_content_generator.py:17]: there is too many logic in this func, please divide some actions to functions

**Review by MichalBor — CHANGES_REQUESTED** [review]
Great job! 👏🏻

File: cycode/cli/code_scanner.py
Comment (avishaiamiel) [reply cycode/cli/code_scanner.py:342]: @MichalBor No, tf plan is not something you are pushing to a repo.
It contains secrets and addition data the injected during build process. 
hence, it's not relevant for repo scan and not implemented in BE.

File: cycode/cli/code_scanner.py
Comment (avishaiamiel) [reply cycode/cli/code_scanner.py:346]: @MichalBor I think that yes, the use case here is that the user will want to run scan over tf plan during pipeline so the expectation is that we will block the pipeline if scan didnt pass successfull.
such case is end case because tf plan is something the automatically generated but if it will happen I believe that it's better to throw error instead os swallow the exception.

File: cycode/cli/code_scanner.py
Comment (avishaiamiel) [reply cycode/cli/code_scanner.py:1124]: @MichalBor My concern is that if terraform_version won't be exists we won't scan the plan even if resource_changes exists.. from other hand, if teffaform_version is always exists it will do the condition stronger which is good. need to think about it a little bit more.

Comment (doratias18) [reply cycode/cli/code_scanner.py:1124]: Need to see if terraform version affects the plan structure (and then to the convert logic)

File: cycode/cli/code_scanner.py
Comment (doratias18) [reply cycode/cli/code_scanner.py:342]: Running terraform plan is only on CI/CD process,  the use-case is that the customer will run it on his build system

File: cycode/cli/code_scanner.py
Comment (MarshalX) [inline cycode/cli/code_scanner.py:344]: mb raise TfplanKeyError inside generate_tf_content_from_tfplan instead of catching KeyError in the root?

File: cycode/cli/code_scanner.py
Comment (MarshalX) [reply cycode/cli/code_scanner.py:346]: looks like it will print the error immediately and will break the output of the progress bar

File: cycode/cli/utils/file_utils.py
Comment (MarshalX) [inline cycode/cli/utils/file_utils.py:6]: f-string instead?

File: cycode/cli/code_scanner.py
Comment (MarshalX) [inline cycode/cli/code_scanner.py:1123]: let's create a function that will load json and handle all errors. if errors have occurred the None will be returned. for example check "get_file_content" function

File: cycode/cli/helpers/tf_content_generator.py
Comment (MarshalX) [inline cycode/cli/helpers/tf_content_generator.py:13]: let's create a function that will load json and handle all errors. if errors have occurred the None will be returned. for example check "get_file_content" function

File: cycode/cli/models.py
Comment (MarshalX) [inline cycode/cli/models.py:68]: use @dataclass or NamedTuple instead to create data models

File: tests/cli/helpers/test_tf_content_generator.py
Comment (MarshalX) [reply tests/cli/helpers/test_tf_content_generator.py:17]: should be possible to reuse "get_file_content" function here to read from files

**Review by MarshalX — COMMENTED** [review]
well done!

File: cycode/cli/code_scanner.py
Comment (MarshalX) [inline cycode/cli/code_scanner.py:340]: we have "perform_pre_scan_documents_actions" function that adds additional files for SCA. I guess it's the best place to add specific logic for IaC. pls check

File: cycode/cli/code_scanner.py
Comment (MarshalX) [inline cycode/cli/code_scanner.py:1126]: `json.JSONDecodeError` instead? to be sure that we are handling json errors properly only and not something too general

File: cycode/cli/utils/file_utils.py
Comment (MarshalX) [reply cycode/cli/utils/file_utils.py:5]: it looks like we don't use "old_ext" var. pls rename it to "_"

File: cycode/cli/helpers/tf_content_generator.py
Comment (EfratIsrael) [reply cycode/cli/helpers/tf_content_generator.py]: We want to stop parsing if the keys are missing, and errors are caught if this is the case

File: cycode/cli/models.py
Comment (MarshalX) [inline cycode/cli/models.py:71]: pls remove constructor 

example:
```python
from dataclasses import dataclass

@dataclass
class InventoryItem:
    """Class for keeping track of an item in inventory."""
    name: str
    unit_price: float
    quantity_on_hand: int = 0

    def total_cost(self) -> float:
        return self.unit_price * self.quantity_on_hand
```

more info: https://docs.python.org/3/library/dataclasses.html

**Review by doratias18 — CHANGES_REQUESTED** [review]
More tests (more tf plans, maybe think on different terraform versions, or at least let's document we support tf version >= X)

File: cycode/cli/code_scanner.py
Comment (EfratIsrael) [reply cycode/cli/code_scanner.py:1112]: #cycode_sast_ignore_here

Comment (EfratIsrael) [reply cycode/cli/code_scanner.py:1112]: #cycode_sast_ignore_here

**Review by doratias18 — CHANGES_REQUESTED** [review]
More tests + add to README how to run a scan on a tf plan (like you wrote in the JIRA)

File: cycode/cli/code_scanner.py
Comment (MarshalX) [reply cycode/cli/code_scanner.py:1119]: #cycode_sast_ignore_here

File: cycode/cli/code_scanner.py
Comment (MarshalX) [reply cycode/cli/code_scanner.py]: fixed in https://github.com/cycodehq-public/cycode-cli/pull/153/commits/7c3296a9867901c3b40a32c1727a620dbbf746be not it doesn't break progress bar

**Review by doratias18 — APPROVED** [review]

**Comment by MarshalX** [comment]
> Improve README so it will look better (@MarshalX maybe you can help)

i think @anna-aleksandrowicz could help with it better than me

File: cycode/cli/code_scanner.py
Comment (EfratIsrael) [reply cycode/cli/code_scanner.py:1119]: #cycode_sast_ignore_here

**Comment by MarshalX** [comment]
@EfratIsrael  pls re-request review when you will be ready

**Review by doratias18 — DISMISSED** [review]
Looks very good now. Thanks for your hard work.

File: README.md
Comment (MarshalX) [inline README.md:427]:
```suggestion
Cycode CLI supports Terraform plan scanning (supporting Terraform 0.12 and later)
```

File: README.md
Comment (MarshalX) [inline README.md:28]:
```suggestion
        1. [Terraform Plan Scan](#terraform-plan-scan)
```

File: README.md
Comment (MarshalX) [inline README.md:426]:
```suggestion
### Terraform Plan Scan

```

File: README.md
Comment (MarshalX) [inline README.md:431]:
```suggestion
```

File: cycode/cli/code_scanner.py
Comment (MarshalX) [inline cycode/cli/code_scanner.py:342]: we already iterate by documents. we can add 

if _is_tfplan_document:
 ... do magic

the thing here instead of creating one more loop below

it will 100% match the logic that tf plans should be only in "scan path" 

it will look like this:
```python
if _is_iac(scan_type) and is_tfplan_document(file, content):
    documents.append(_handle_tflpan_document(...))
else:
    documents.append(Document(file, content, is_git_diff))
```

File: cycode/cli/helpers/tf_content_generator.py
Comment (MarshalX) [inline cycode/cli/helpers/tf_content_generator.py:8]: list of lists looks complicated :(

File: cycode/cli/code_scanner.py
Comment (MarshalX) [inline cycode/cli/code_scanner.py:1136]: load_json could return None. we should check that value not != None. otherwise, the code will fail below because of "TypeError: argument of type 'NoneType' is not iterable"

File: cycode/cli/helpers/tf_content_generator.py
Comment (MarshalX) [inline cycode/cli/helpers/tf_content_generator.py:34]: we should care about `None` value here too. I understand that checks occur in the function in the root method (`_is_tfplan_document`), but it's not clear. if someone in the future will use this function without `_is_tfplan_document` before it will raise the exception (`TypeError: 'NoneType' object is not subscriptable
`).

File: cycode/cli/helpers/tf_content_generator.py
Comment (MarshalX) [inline cycode/cli/helpers/tf_content_generator.py:38]: possible exception `TypeError: 'NoneType' object is not iterable` is not handled. it will also occur when `_is_tfplan_document` was not called before. 

maybe `tfplan_json.get('resource_changes', [])`

or  `except (KeyError, TypeError) as e:`

**Review by MarshalX — CHANGES_REQUESTED** [review]
good structure of test files! i like that! 

btw do we expect that tf plans files could be in the directory with another IaC related files? "scan path" will scan everything related to IaC. 

im asking because i wonder that our user-firendly exception doesn't show the path to invalid tf plan file. will it be a problem for the user to find it in the provided path with many subfolders?

File: cycode/cli/helpers/tf_content_generator.py
Comment (EfratIsrael) [reply cycode/cli/helpers/tf_content_generator.py:8]: not relevant anymore

File: cycode/cli/helpers/tf_content_generator.py
Comment (EfratIsrael) [reply cycode/cli/helpers/tf_content_generator.py:38]: used this `except (KeyError, TypeError) as e` because we want to stop generating if `resource_changes` missing at this point

File: cycode/cli/code_scanner.py
Comment (EfratIsrael) [reply cycode/cli/code_scanner.py]: Terraform 0.12 and later is supported

File: cycode/cli/code_scanner.py
Comment (EfratIsrael) [reply cycode/cli/code_scanner.py:1136]: I'll use .get with False fallback

File: cycode/cli/code_scanner.py
Comment (EfratIsrael) [reply cycode/cli/code_scanner.py:342]: yes, I'll ad something similar to this suggestion

File: cycode/cli/helpers/tf_content_generator.py
Comment (EfratIsrael) [reply cycode/cli/helpers/tf_content_generator.py:34]: I think it is covered with the typeError handling afterwards. this is in the scope of generate_tf_content_from_tfplan so it suppose to be json and have resource_changes and if not it will be caught in the typeError

**Review by MichalBor — DISMISSED** [review]
Amazing ⭐ 
please wait for Ilya approval

File: cycode/cli/code_scanner.py
Comment (MarshalX) [reply cycode/cli/code_scanner.py]: but now we have `AttributeError: 'NoneType' object has no attribute 'get'` 😢

```
tf_plan = load_json(document.content)
if tf_plan is None:
    return False

return 'resource_changes' in tf_plan
```

**Review by doratias18 — DISMISSED** [review]

File: tests/cli/test_code_scanner.py
Comment (EfratIsrael) [reply tests/cli/test_code_scanner.py:138]: #cycode_sast_ignore_here

File: tests/cli/test_code_scanner.py
Comment (EfratIsrael) [reply tests/cli/test_code_scanner.py:111]: #cycode_sast_ignore_here

File: tests/cli/test_code_scanner.py
Comment (EfratIsrael) [reply tests/cli/test_code_scanner.py:95]: #cycode_sast_ignore_here

**Review by MarshalX — COMMENTED** [review]
awesome!

**Review by MarshalX — APPROVED** [review]
rly hard work!

**Pull request merged**
